### PR TITLE
Replace ProjectGenerator

### PR DESCRIPTION
**Pull request by giginet**
Since Swift Package Manager 5.8, Xcodeproj generation is obsoleted.
https://github.com/apple/swift-package-manager/pull/5748

I ported this implementation. I'll add some tests.
Use `tuist/XcodeProj` to manipulate Xcode projects instead.

Screen Shot 2022-11-15 at 19 21 37

## Current limitation

- Current implementation supports resources partially only
    - Can't use `Bundle.module` frameworks built with Scipio. It will be supported on #11 
- Not well tested
    - I'll add tests for ClangTarget in another PR

### CODE REVIEW

File: Sources/ScipioKit/Compiler.swift [hunk 1]
```diff
@@ -76,7 +76,6 @@ struct Compiler<E: Executor> {
                             try fileSystem.removeFileTree(at: xcframeworkPath)
                         }
                     }
-                    try fileSystem.removeFileTree(at: xcframeworkPath)
```
Comment (giginet) [inline Sources/ScipioKit/Compiler.swift:79]: It is called twice...

File: Package.swift [hunk 2]
```diff
@@ -16,14 +16,18 @@ let package = Package(
             targets: ["ScipioKit"]),
     ],
     dependencies: [
+        // A revision after Xcodeproj module is removed
+        // https://github.com/apple/swift-package-manager/pull/5748
         .package(url: "https://github.com/apple/swift-package-manager.git",
-                 revision: "swift-5.7-RELEASE"),
```
Comment (giginet) [inline Package.swift:20]: swift-package-manager on `swift-5.7-RELEASE` contains `Xcodeproj` module.
Unfortunately, its conflicts with `tuist/XcodeProj`.

So I pined a version to the latest.

@freddi-kit

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+                    collection[target] = moduleMapPath
+                }
+            }
+        }
+
+        // Make LinkPhase for each Xcode targets
+        for (target, xcodeTarget) in xcodeTargets {
+            let dependsTargets = try target.recursiveDependencies().compactMap { value in
+                if case .target(let dependency, _) = value {
+                    return dependency
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:159]: `value` is ambiguous.
How about this change?
```suggestion
            let dependsTargets = try target.recursiveDependencies().compactMap { dependency in
                if case .target(let resolvedTarget, _) = dependency {
                    return resolvedTarget
```

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+        if let productGroup = pbxProj.rootObject?.productsGroup {
+            productRef = try productGroup.addFile(
+                at: target.productPath.toPath(),
+                sourceTree: .buildProductsDir,
+                sourceRoot: target.sources.root.toPath(),
+                validatePresence: false
+            )
+        } else {
+            productRef = nil
+        }
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:234]: The if statement can be removed.
```suggestion
        let productRef: PBXFileReference? = try? pbxProj.rootObject?
            .productsGroup?
            .addFile(
                at: target.productPath.toPath(),
                sourceTree: .buildProductsDir,
                sourceRoot: target.sources.root.toPath(),
                validatePresence: false
            )
```

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+
+        guard let sourceRoot = sourceRoot else {
+            throw Error.unknownError
+        }
+
+        guard let mainGroup = pbxProj.rootObject?.mainGroup else {
+            throw Error.unknownError
+        }
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:242]: The two guard statements can be merged.
```suggestion
        guard let sourceRoot = sourceRoot,
              let mainGroup = pbxProj.rootObject?.mainGroup else {
            throw Error.unknownError
        }
```

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+            )
+            return try group.addFile(
+                at: sourcePath.toPath(),
+                sourceTree: .sourceRoot,
+                sourceRoot: sourceRoot.toPath()
+            )
+        }
+
+        let buildFiles: [PBXBuildFile] = fileReferences.map { reference in
+            return addObject(PBXBuildFile(file: reference))
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:260]:
```suggestion
            addObject(PBXBuildFile(file: reference))
```

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+        }
+
+        let compilePhase = addObject(
+            PBXSourcesBuildPhase(
+                files: buildFiles
+            )
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:266]:
```suggestion
            PBXSourcesBuildPhase(files: buildFiles)
```

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+            let resourcesGroup = try targetGroup.addGroup(named: "Resources").first!
+
+            switch resource.rule {
+            case .process(let localization):
+                guard localization == nil else {
+                    throw Error.localizationNotSupported(targetName: target.name)
+                }
+            case .copy:
+                break
+            }
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:306]: If statement can represent your intension more.
```suggestion
            if case let .process(localization) = resource.rule,
               localization != nil {
                throw Error.localizationNotSupported(targetName: target.name)
            }
```

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+
+            let files = try walk(resource.path)
+            return try files.map { file in
+                return try resourcesGroup.addFile(at: file.toPath(), sourceRoot: sourceRoot.toPath())
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:310]:
```suggestion
                try resourcesGroup.addFile(at: file.toPath(), sourceRoot: sourceRoot.toPath())
```

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+        }
+
+        let buildFiles = resourcesReferences.map { PBXBuildFile(file: $0) }
+
+        return PBXResourcesBuildPhase(files: buildFiles)
+    }
+
+    private struct ClangTargetSettingResult {
+        var moduleMapPath: AbsolutePath?
+    }
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:321]: Is this struct really required?

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+        for header in headerFiles {
+            let headerFileGroup = try self.group(
+                for: header.parentDirectory,
+                parentGroup: targetGroup,
+                sourceRoot: clangTarget.sources.root
+            )
+            let fileRef = try headerFileGroup.addFile(at: header.toPath(),
+                                                      sourceRoot: clangTarget.path.toPath())
+            headerFileRefs.append(fileRef)
+        }
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:341]: `headerFileRefs` can become `let`
```suggestion
        let headerFileRefs: [PBXFileReference] = try headerFiles.map { header in
            let headerFileGroup = try self.group(
                for: header.parentDirectory,
                parentGroup: targetGroup,
                sourceRoot: clangTarget.sources.root
            )
            return try headerFileGroup.addFile(at: header.toPath(), sourceRoot: clangTarget.path.toPath())
        }
```

**Review by mtj0928 — COMMENTED** [review]
I don't have good knowledge about build system, so I reviewed this PR from the viewpoint of readability.

File: Sources/ScipioKit/ProjectGenerator/BuildSettingsGenerator.swift [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+            "GCC_PREPROCESSOR_DEFINITIONS": [.inherited, "SWIFT_PACKAGE=1"],
+            "USE_HEADERMAP": false,
+            "CLANG_ENABLE_OBJC_ARC": true,
+        ]
+        return values.mapValues(\.rawValue)
+    }
+
+    private var debugSpecificSettings: BuildSettings {
+        let specificSettings: [String: XCConfigValue] = [
+            "COPY_PHASE_STRIP": false,
```
Comment (freddi-kit) [inline Sources/ScipioKit/ProjectGenerator/BuildSettingsGenerator.swift:85]: Why don't we necessary TARGETED_DEVICE_FAMILY?

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift [hunk 13]
```diff
@@ -0,0 +1,505 @@
...
+        xcodeTargets.values.forEach { self.pbxProj.rootObject?.targets.append($0) }
+
+        // Generate ModuleMaps for Clang targets
+        let moduleMaps: [ResolvedTarget: AbsolutePath] = try xcodeTargets.reduce(into: [:]) { (collection, tuple) in
+            let (target, xcodeTarget) = tuple
+            if let clangTarget = target.underlyingTarget as? ClangTarget {
+                if let moduleMapPath = try applyClangTargetSpecificSettings(for: clangTarget, xcodeTarget: xcodeTarget).moduleMapPath {
+                    collection[target] = moduleMapPath
+                }
+            }
```
Comment (mtj0928) [inline Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:152]:
```suggestion
            if let clangTarget = target.underlyingTarget as? ClangTarget,
               let moduleMapPath = try applyClangTargetSpecificSettings(for: clangTarget, xcodeTarget: xcodeTarget).moduleMapPath {
                    collection[target] = moduleMapPath
                }
            }
```

File: Sources/ScipioKit/ProjectGenerator/BuildSettingsGenerator.swift [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+import XcodeProj
+
+enum XCConfigValue {
+    case string(String)
+    case list([XCConfigValue])
+    case bool(Bool)
+
+    static let inherited: Self = .string("$(inherited)")
+
+    var rawValue: Any {
```
[the remark this answers is not in the document]
Comment (giginet) [reply Sources/ScipioKit/ProjectGenerator/BuildSettingsGenerator.swift:17]: remame to `rawConfigValue`

File: Sources/ScipioKit/BuildOptions.swift [hunk 15]
```diff
@@ -1,6 +1,20 @@
...
+        frameworkType: FrameworkType,
+        sdks: [SDK]
+    ) {
+        self.buildConfiguration = buildConfiguration
+        self.isSimulatorSupported = isSimulatorSupported
+        self.isDebugSymbolsEmbedded = isDebugSymbolsEmbedded
+        self.frameworkType = frameworkType
+        self.sdks = sdks
+    }
+
```
[the remark this answers is not in the document]
Comment (giginet) [reply Sources/ScipioKit/BuildOptions.swift:17]: I'm using this constructor from tests

File: Sources/ScipioKit/BuildOptions.swift [hunk 16]
```diff
@@ -1,6 +1,20 @@
...
    var buildConfiguration: BuildConfiguration
```
[the remark this answers is not in the document]
Comment (giginet) [reply Sources/ScipioKit/BuildOptions.swift:18]: oops

File: Sources/ScipioKit/ProjectGenerator/BuildSettingsGenerator.swift [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+            "COMBINE_HIDPI_IMAGES": true,
+            "SWIFT_ACTIVE_COMPILATION_CONDITIONS": [.inherited, "SWIFT_PACKAGE"],
+            "GCC_PREPROCESSOR_DEFINITIONS": [.inherited, "SWIFT_PACKAGE=1"],
+            "USE_HEADERMAP": false,
+            "CLANG_ENABLE_OBJC_ARC": true,
+        ]
+        return values.mapValues(\.rawValue)
+    }
+
+    private var debugSpecificSettings: BuildSettings {
```
[the remark this answers is not in the document]
Comment (giginet) [reply Sources/ScipioKit/ProjectGenerator/BuildSettingsGenerator.swift:83]: Yes! I referred to most parts from the original implementation https://github.com/apple/swift-package-manager/pull/5748

File: Sources/ScipioKit/XcodeBuild/Commands/ArchiveCommand.swift [hunk 18]
```diff
@@ -14,7 +14,7 @@ struct ArchiveCommand: XcodeBuildCommand {
         [
             ("project", projectPath.path),
             ("configuration", buildConfiguration.settingsValue),
-            ("scheme", target.name),
+            ("scheme", target.c99name),
```
[the remark this answers is not in the document]
Comment (giginet) [reply Sources/ScipioKit/XcodeBuild/Commands/ArchiveCommand.swift:17]: some packages contains invalid characters for target names.

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift (hunk 11 above)
Comment (giginet) [reply Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:340]: I don't like to have side-effects inside `map` a little....

File: Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift (hunk 10 above)
Comment (giginet) [reply Sources/ScipioKit/ProjectGenerator/ProjectGenerator.swift:320]: I feel the return value of `applyClangTargetSpecificSettings` seems to be ambiguous.
So I wrap `AbsolutePath` in a named struct.

File: Sources/ScipioKit/ProjectGenerator/BuildSettingsGenerator.swift [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+            "OTHER_SWIFT_FLAGS": [.inherited],
+            "SWIFT_ACTIVE_COMPILATION_CONDITIONS": [.inherited],
+            "FRAMEWORK_SEARCH_PATHS": [.inherited, "$(PLATFORM_DIR)/Developer/Library/Frameworks"],
+            "BUILD_LIBRARY_FOR_DISTRIBUTION": true,
+        ]
+    }
+
+    func generate(for target: ResolvedTarget, configuration: BuildConfiguration, infoPlistPath: URL) -> XCBuildConfiguration {
+        var settings: [String: XCConfigValue] = baseSettings(for: target)
+        settings["INFOPLIST_FILE"] = .string(infoPlistPath.relativePath)
```
Comment (freddi-kit) [inline Sources/ScipioKit/ProjectGenerator/BuildSettingsGenerator.swift:141]:
```suggestion
        let path = infoPlistPath.relativeString.components(separatedBy: "/")
            .dropFirst(1)
            .joined(separator: "/")
        settings["INFOPLIST_FILE"] = .string(path)
```

Comment (giginet) [reply Sources/ScipioKit/ProjectGenerator/BuildSettingsGenerator.swift]: This change seems to be not correct

**Review by freddi-kit — APPROVED** [review]
Good job

**Review by mtj0928 — APPROVED** [review]

**Pull request merged**